cloudflare/resource_cloudflare_rate_limit.go

```diff
@@ -0,0 +1,490 @@
+			for i, s := range schemesSet.(*schema.Set).List() {
+				schemes[i] = s.(string)
+			}
+			requestMatcher.Schemes = schemes
+		}
+		matcher.Request = requestMatcher
+	}
+	if matchRespIface, ok := cfg["response"]; ok && len(matchRespIface.([]interface{})) > 0 {
+		matchResp := matchRespIface.([]interface{})[0].(map[string]interface{})
+
+		responseMatcher := cloudflare.RateLimitResponseMatcher{}
+
+		if statusesSet, ok := matchResp["statuses"]; ok {
+			statuses := make([]int, statusesSet.(*schema.Set).Len())
+			for i, s := range statusesSet.(*schema.Set).List() {
+				statuses[i] = s.(int)
+			}
+			responseMatcher.Statuses = statuses
+		}
+		if originIface, ok := matchResp["origin_traffic"]; ok {
+			originTraffic := originIface.(bool)
+			responseMatcher.OriginTraffic = &originTraffic
+		}
+		matcher.Response = responseMatcher
+	}
+	return
+}
+
+func expandRateLimitAction(d *schema.ResourceData) cloudflare.RateLimitAction {
+	// dont need to guard for array length because MinItems is set **and** action is required
+	tfAction := d.Get("action").([]interface{})[0].(map[string]interface{})
+
+	action := cloudflare.RateLimitAction{
+		Mode:    tfAction["mode"].(string),
+		Timeout: tfAction["timeout"].(int),
+	}
+
+	if _, ok := tfAction["response"]; ok && len(tfAction["response"].([]interface{})) > 0 {
+		log.Printf("[DEBUG] CloudFlare Rate Limit specified action: %+v \n", tfAction)
+		tfActionResponse := tfAction["response"].([]interface{})[0].(map[string]interface{})
+
+		action.Response = &cloudflare.RateLimitActionResponse{
+			ContentType: tfActionResponse["content_type"].(string),
+			Body:        tfActionResponse["body"].(string),
+		}
+	}
+	return action
+}
+
+func expandRateLimitBypass(bypassUrlPatterns *schema.Set) []cloudflare.RateLimitKeyValue {
+	bypass := make([]cloudflare.RateLimitKeyValue, bypassUrlPatterns.Len())
+	for i, urlPattern := range bypassUrlPatterns.List() {
+		bypass[i] = cloudflare.RateLimitKeyValue{
+			Name:  "url",
+			Value: urlPattern.(string),
+		}
+	}
+	return bypass
+}
+
+func resourceCloudFlareRateLimitRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*cloudflare.API)
+	zoneId := d.Get("zone_id").(string)
+	rateLimitId := d.Get("rate_limit_id").(string)
+
+	rateLimit, err := client.RateLimit(zoneId, rateLimitId)
+	if err != nil {
+		if strings.Contains(err.Error(), "HTTP status 404") {
+			log.Printf("[INFO] Resource %s in zone %s no longer exists", rateLimitId, zoneId)
+			d.SetId("")
+			return nil
+		} else {
+			return errors.Wrap(err,
+				fmt.Sprintf("Error reading rate limit resource from API for resource %s in zone %s", zoneId, rateLimitId))
+		}
+	}
+	log.Printf("[DEBUG] Read CloudFlare Rate Limit from API as struct: %+v", rateLimit)
+
+	d.Set("threshold", rateLimit.Threshold)
+	d.Set("period", rateLimit.Period)
+	d.Set("match", flattenRateLimitTrafficMatcher(rateLimit.Match))
+	d.Set("action", flattenRateLimitAction(rateLimit.Action))
```

Comment:
Errors need to be checked when setting non-scalar attributes such as `action` and `match`:

```go
if err := d.Set("action", ...); err != nil {
  log.Printf("[WARN] Error setting action: %s", err)
}
```